BootloaderCorePkg/Library/S3SaveRestoreLib/S3SaveRestore.c

```diff
@@ -0,0 +1,162 @@
+/** @file
+
+  Copyright (c) 2019, Intel Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <PiPei.h>
+#include <Library/DebugLib.h>
+#include <Library/HobLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/S3SaveRestoreLib.h>
+#include <Library/IoLib.h>
+#include <Guid/SmmInformationGuid.h>
+#include <Library/BoardInitLib.h>
+#include <Guid/SmmInformationGuid.h>
+
+
+/**
+  This function uploads information in TSEG area
+  designated for S3 save/restore purpose.
+
+  @param    DataPtr       Address of the structure to be copied to TSEG
+
+  @retval   None
+**/
+VOID
+EFIAPI
+SetS3Info (
+  IN  VOID    *DataPtr
+  )
+{
+  UINT8             *Ptr;
+  LDR_SMM_INFO      LdrSmmInfo;
+  UINT8             *SmmBase;
+  UINT32            SmmSize;
+
+  PlatformUpdateHobInfo (&gSmmInformationGuid, &LdrSmmInfo);
+  SmmBase = (UINT8 *) LdrSmmInfo.SmmBase;
+  SmmSize = LdrSmmInfo.SmmSize;
```

Comment:
The SmmSize here is the whole SMM range size.
need check if flag SMM_FLAGS_4KB_COMMUNICATION is set in LdrSmmInfo.Flags. if yes. set SmmSize to 4KB. else return.

Reply:
Good catch. Thanks, will fix it.